### ISSUE TITLE
Make it possible to reset recorded calls on an object

### ISSUE THREAD

**Issue by erik-kallen** [FakeItEasy/FakeItEasy#978]
It is sometimes useful (especially for integration testing) to have a non-trivial setup phase that might call methods on fakes, but then you want to exclude those calls from the assertions.

I propose adding the API

```c#
Fake.ClearRecordedCalls(instance);
```

So you can do things like

```c#
public void MyTest() {
    // Arrange
    var service = A.Fake<IMyService>();
    var sut = new MyClass(service);
    sut.DoSomeInitialization();
    Fake.ClearRecordedCalls(service);

    // Act
    sut.DoTheStuffThatShouldBeTested();

    // Assert
    A.CallTo(() => service.SomeServiceMethod(1, 2)).MustHaveHappened); // Ensures that this call has happened during the act phase
}
```

This used to be doable with scopes in 1.x but I can't find any way to achieve my goal in 2.x

**Comment by blairconrad** [comment]
Hi, @erik-kallen. Thanks for your interest.

Everyone works differently, but this seems like a peculiar workflow. A similar question was raised a couple of years ago on StackOverflow, and I agree with [forsvarir's answer](http://stackoverflow.com/questions/29095019/fakeiteasy-reset-fake-calls-history-ignore-call/29096765#29096765), which I reproduce here:

----

 This feels wrong to me.  Typically I would expect the cost of object creation to be part of the expectations of the tests for that object.  Obviously, it depends upon what your `AssertedMethod` does, however, would your test really still be valid if your constructor didn’t call it?  Would your SUT be in the expected state that your class needs?

I can’t see anything in the fakeiteasy framework to allow you to reset the count, although as I’ve said, this doesn’t feel like the right thing to do so I wasn’t really expecting to find it.  It seems like a simple way to achieve what you seem to be trying to do, whilst still acknowledging at least at a `TestClass` level that you know there is a call in the constructor/init would be something like this:

```c#
const int InitAssertedMethodCalls = 1;

[TestMethod]
public void TestMethod1()
{
    sut.TestedMethod(); 
    A.CallTo(() => myFakedObject.AssertedMethod())
             .MustHaveHappened(Repeated.Exactly.Times(1 + InitAssertedMethodCalls )); 
}
```

**Comment by erik-kallen** [comment]
I want to exclude the setup phase from my assertions because I don't want the tests to break if the setup somehow changes (as I said, this is for integration testing).

The proposed solution only works if there are no arguments to the call, but one of the best features with FakeItEasy is that it is very easy to assert argument values.

**Comment by blairconrad** [comment]
I see your point, @erik-kallen. It still feels a little weird to me, but not everyone likes to work the same way.
And really, it doesn't feel that different than  `Fake.ClearConfiguration` (which I'm not a fan of either, but there you go). Maybe it's even a better method to have in the API than `ClearConfiguration`.

@thomaslevesque, @adamralph, what do you think?

----
(aside)
For a second there, I thought there'd be a way to accomplish what you want by wrapping another fake, but I think it ends up being too much of a pain. The sketch of the idea is to make an object of your faked type. This object could be switchable between two different backing fakes. You construct your SUT with this switchable fake, which is configured to point at a FakeItEasy-generated fake. Then after initialization, you configure your switchable fake to point to a whole new fake, which wraps the first inner fake. Then you assert on that second inner fake.
I think it's achievable. But sort of gross.

**Comment by erik-kallen** [comment]
I discovered that it is, indeed, possible to do what I want to, but it is rather hacky:

            var queue = (ConcurrentQueue<ICompletedFakeObjectCall>)Fake.GetCalls(instance);
            ICompletedFakeObjectCall call;
            while (queue.TryDequeue(out call))
            {
            }

**Comment by thomaslevesque** [comment]
It seems a bit strange to me too; I can't really think of a scenario where you would need to make actual calls to a fake to configure it. But I don't have a clear picture of @erik-kallen's situation, so I might be missing something.
Anyway, I'm not opposed to a `ClearConfiguration` method; it's easy to implement and doesn't do any harm, as far as I can tell.

**Comment by blairconrad** [comment]
@erik-kallen, that is clever! And something we should probably have prevented by returning a copy of the calls.

Here's my EVEN MORE TERRIBLE solution:

```c#
public class ResetCountsAfterSetupExample
{
    public interface IFoo
    {
        void Method(int i);
    }

    public class FooSwitcher : IFoo
    {
        private IFoo innerFake;

        public void SetInnerFake(IFoo fake)
        {
            innerFake = fake;
        }

        public void Method(int i)
        {
            innerFake.Method(i);
        }
    }

    public class Sut
    {
        private readonly IFoo foo;

        /// <summary>
        /// Initializes a new instance of the <see cref="T:System.Object"/> class.
        /// </summary>
        public Sut(IFoo foo)
        {
            this.foo = foo;
        }

        public void DoSomething(int i)
        {
            foo.Method(i);
        }
    }

    [Test]
    public void T()
    {
        var initialzationFoo = A.Fake<IFoo>();
        var testingFoo = A.Fake<IFoo>(options => options.Wrapping(initialzationFoo));

        var fooSwitcher = new FooSwitcher();
        fooSwitcher.SetInnerFake(initialzationFoo);

        var sut = new Sut(fooSwitcher);

        // Very fancy initialization
        sut.DoSomething(1);

        // get ready for the test
        fooSwitcher.SetInnerFake(testingFoo);

        // the actual test
        sut.DoSomething(17);

        A.CallTo(() => testingFoo.Method(17)).MustHaveHappened(Repeated.Exactly.Once);
        A.CallTo(() => testingFoo.Method(1)).MustNotHaveHappened();
    }
}
```

**Comment by erik-kallen** [comment]
@thomaslevesque I'm not making actual calls to a fake to configure it. I'm making call to a service that uses the fake as a dependency, and those calls might call the fake but I wish to disregard those calls when performing my asserts.

**Comment by thomaslevesque** [comment]
> @thomaslevesque I'm not making actual calls to a fake to configure it. I'm making call to a service that uses the fake as a dependency, and those calls might call the fake but I wish to disregard those calls when performing my asserts.

Ah, I see. Thanks for the clarification.

**Comment by blairconrad** [comment]
@erik-kallen, as this seems near and dear to your heart, would you be interested in implementing it?

**Comment by thomaslevesque** [comment]
> Anyway, I'm not opposed to a `ClearConfiguration` method; it's easy to implement and doesn't do any harm, as far as I can tell.

I meant `ClearRecordedCalls`, of course... `ClearConfiguration` is something else entirely, which isn't really useful IMO.

**Comment by blairconrad** [comment]
@erik-kallen is not interested in continuing, so I'm taking over.

**Comment by blairconrad** [comment]
This issue was fixed in [Release 3.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.2.0).

@erik-kallen, look for your name in the release notes! :trophy:
